### ISSUE TITLE
balloons in the terminal only work with balloon_show and only asynchronously

### ISSUE THREAD

**Issue by puremourning** [vim/vim#2481]
# Overview

When using a balloons in the terminal, a simple `balloonexpr` which returns a message does not work. It only displays the balloon/message (curiously) when hitting `:` to enter command line mode.

If a _subsequent_ (async) call to `balloon_show()` is made, the balloon appears as expected.

# Steps to reproduce

* Save the following as testballoon.vim

```vim
let s:counter = 0
let s:timer = -1

function! MyBalloonExpr()
  let s:counter += 1
  call timer_stop( s:timer )
  let s:message =
          \ 'Cursor is at line ' . v:beval_lnum .
          \', column ' . v:beval_col .
          \ ' of file ' .  bufname(v:beval_bufnr) .
          \ ' on word "' . v:beval_text . '"'
  return s:message
endfunction

set mouse=a
set ttymouse=sgr
set balloonexpr=MyBalloonExpr()
set balloondelay=250
set ballooneval
set balloonevalterm
```

* run command `vim -Nu NONE -S testballoon.vim testballoon.vim`
* hover mouse on the word `MyBalloonExpr`
* EXPECT: The balloon shows up after 250ms
* ACTUAL: No balloon
* Hit `:` (to enter command line)
* EXPECT: No balloon?
* ACTUAL: Balloon is shown, but with possibly garbled message.

# Alternative (working) implementation

To demonstrate the point about async:

* Save the following as testballoon.vim

```vim
let s:counter = 0
let s:timer = -1

function! MyBalloonExpr()
  let s:counter += 1
  call timer_stop( s:timer )
  let s:message =
          \ 'Cursor is at line ' . v:beval_lnum .
          \', column ' . v:beval_col .
          \ ' of file ' .  bufname(v:beval_bufnr) .
          \ ' on word "' . v:beval_text . '"'
  let s:timer = timer_start( 0, 'RealBalloonExpr' )
endfunction

function! RealBalloonExpr(timer)
  echom 'MyBalloonExpr: ' . s:counter
  call balloon_show( s:message )
endfunction

set mouse=a
set ttymouse=sgr
set balloonexpr=MyBalloonExpr()
set balloondelay=250
set ballooneval
set balloonevalterm
```

* run command `vim -Nu NONE -S testballoon.vim testballoon.vim`
* hover mouse on the word `MyBalloonExpr`
* EXPECT: The balloon shows up after 250ms
* ACTUAL: The balloon shows correctly after 250ms
* Hit `:` (to enter command line)
* EXPECT: Balloon persists
* ACTUAL: Balloon persists

# Secondary/related issue ?

In `popupmenu.c`, it doesn't look like `balloon_mouse_row` or `balloon_mouse_col` are ever set, but they are read by `ui_may_remove_balloon()`.

**Comment by puremourning** [comment]
Maybe tangentially related to https://github.com/vim/vim/issues/2352 (or might at least explain _some_ of the confusion?)

**Comment by sam-mccall** [comment]
Very strangely, I see the balloon with the return value of balloonexpr, but only if balloonexpr sets a timer.

This works (shows 99):
```
function BalloonExpr()
  call timer_start(0, {t->0})
  return 99
endfunction

set nocompatible
set mouse=a
set balloonexpr=BalloonExpr()
set bevalterm
```

If the `timer_start` is removed it fails.

EDIT: this also makes subsequent calls to balloon_show work, even if they don't come from the callback. So for anyone who finds this bug, setting a null timer works around the balloonexpr bug, and setting balloonexpr to a function that sets a null timer and returns "" makes show_balloon work everywhere.

**Comment by chrisbra** [comment]
perhaps solved by [8.1.1300](2f10658b06bbdd8f25c4ff15226)?

**Comment by brammool** [comment]
Yes, this should be solved now.  If not, please reopen or file another issue.
